pkg/kvstore/etcd.go

```diff
@@ -544,8 +544,21 @@ func (e *etcdClient) renewSession(ctx context.Context) error {
 }
 
 func (e *etcdClient) renewLockSession(ctx context.Context) error {
-	<-e.firstSession
-	<-e.lockSession.Done()
+	select {
+	// wait for initial session to be established
+	case <-e.firstSession:
+	// controller has stopped or etcd client is closing
+	case <-ctx.Done():
+		return nil
+	}
+
+	select {
+	// session has ended
+	case <-e.session.Done():
+	// controller has stopped or etcd client is closing
+	case <-ctx.Done():
+		return nil
```

Comment:
Shouldn't be `return ctx.Err()`?

Reply:
This is still returning `nil` after all changes unlike the previous case. The code is choosing to not return an error to not trigger a retry when the controller stops or the etcd client shuts down.